### ISSUE TITLE
Persistance storage - docker image

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#4302]
Currently the docker image built for base egeria does not provide any documentation, or direct support, for using persistent storage. For example the configuration files used to configure servers, audit logs , and of course a graph metadata store.

Thus far our docker image has been used for a tutorial environment, where this lack of persistency is actually a good thing .... 
Additionally, egeria itself provides lots of options through connectors & configuration which might allow any storage to be externalized.

However for many simple (including learning, sample) environments it's necessary to support this in a simple way. At a docker level this might mean via a volume. In k8s we would create Persistent Volume Claims.  Further it should be clear in the documentation how this is most easily done.

**Comment by planetf1** [comment]
The broad strategy here is to
 - prove how to configure egeria (and fix if required) so that all data files are in well defined directories that could be mapped to an external storage medium
 - ensure this is documented in our admin guide
 - Extend our docker image accordingly, plus docs
 - Extend our compose, charts accordingly

Detailed todos:
 - [ ] Extend coco pharma lab notebook to create all configuration in a well defined subdirectory (not root)
 - [ ] Extend notebook to support use of graph store instead of inmemory
 - [ ] Ensure graph store is also in a well defined subdirectory
 - [ ] Ensure admin guide is correct & helpful for end users wishing to make this kind of configuration
 - [ ] Extend docker image to easily support/define volumes to be used
 - [ ] Ensure docker usage docs updated
 - [ ] Extend lab helm chart to optionally allow persistant storage to be used

Further deliverables will follow - such as persistance fot the egeria operator, or other deployment options, this is a first step to provide the approach.

**Comment by planetf1** [comment]
Coco pharma notebook now stores files under 'config' by default, making use of the API to set the configuration store
This provides a good example, and validation

ie:
```
jonesn:egeria/ (issue4302) $ ls -la config                                                                                             [15:25:39]
total 624
drwxr-xr-x  11 jonesn  staff    352 14 Dec 15:25 .
drwxr-xr-x  57 jonesn  staff   1824 14 Dec 15:25 ..
-rw-r--r--   1 jonesn  staff  45115 14 Dec 15:25 omag.server.cocoMDS1.config
-rw-r--r--   1 jonesn  staff  51982 14 Dec 15:25 omag.server.cocoMDS2.config
-rw-r--r--   1 jonesn  staff  41319 14 Dec 15:25 omag.server.cocoMDS3.config
-rw-r--r--   1 jonesn  staff  26019 14 Dec 15:25 omag.server.cocoMDS4.config
-rw-r--r--   1 jonesn  staff   7985 14 Dec 15:25 omag.server.cocoMDS5.config
-rw-r--r--   1 jonesn  staff  59899 14 Dec 15:25 omag.server.cocoMDS6.config
-rw-r--r--   1 jonesn  staff  56882 14 Dec 15:25 omag.server.cocoMDSx.config
-rw-r--r--   1 jonesn  staff  10568 14 Dec 15:25 omag.server.cocoView1.config
-rw-r--r--   1 jonesn  staff   3448 14 Dec 15:25 omag.server.findItDL01.config
```

**Comment by planetf1** [comment]
Added environment variable to allow configuration of graph repo in notebooks by default (for testing)

**Comment by planetf1** [comment]
A fix for #1102 is a prereq to this change. The experimental changes to the notebook are not needed once this is done.

**Comment by planetf1** [comment]
#1102 is now submitted as a PR. There we map all dynamic data for egeria into the ./data subdirectory.
This is what we will map in a docker volume or k8s pvc

**Comment by planetf1** [comment]
docker-compose now supports persistence - notes added to readme
odpi-egeria-lab chart - support added. testing 

This is now what happens after deployment of the chart - if persistence is enabled (default):

```
jonesn:charts/ (issue4302b) $ kubectl get pvc                                                                                          [18:33:26]
NAME                                               STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                 AGE
data-lab-kafka-0                                   Pending                                      ibmc-vpc-block-10iops-tier   4m14s
data-lab-zookeeper-0                               Pending                                      ibmc-vpc-block-10iops-tier   4m14s
lab-core-data-lab-odpi-egeria-lab-core-0           Pending                                      ibmc-vpc-block-10iops-tier   4m14s
lab-datalake-data-lab-odpi-egeria-lab-datalake-0   Pending                                      ibmc-vpc-block-10iops-tier   4m14s
lab-dev-data-lab-odpi-egeria-lab-dev-0             Pending                                      ibmc-vpc-block-10iops-tier   4m14s
lab-factory-data-lab-odpi-egeria-lab-factory-0     Pending                                      ibmc-vpc-block-10iops-tier   4m14s
```

When we look at one of these we see:

```
jonesn:charts/ (issue4302b) $ kubectl describe pvc/data-lab-kafka-0                                                                    [18:31:57]
Name:          data-lab-kafka-0
Namespace:     egeria-lab
StorageClass:  ibmc-vpc-block-10iops-tier
Status:        Pending
Volume:
Labels:        app.kubernetes.io/component=kafka
               app.kubernetes.io/instance=lab
               app.kubernetes.io/name=kafka
Annotations:   volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
VolumeMode:    Filesystem
Used By:       lab-kafka-0
Events:
  Type    Reason                Age                 From                         Message
  ----    ------                ----                ----                         -------
  Normal  ExternalProvisioning  7s (x18 over 4m9s)  persistentvolume-controller  waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  ```
  If we do a describe on one of the stateful sets (needed changing from deployment to accommodate the volumes) ie `kubectl describe statefulset/lab-odpi-egeria-lab-core` we see in the output:
  ```
      Mounts:
      /deployments/data from lab-core-data (rw)
   ```
   
   and correctly in the event log:
   ```
     Normal  SuccessfulCreate  5m22s  statefulset-controller  create Claim lab-core-data-lab-odpi-egeria-lab-core-0 Pod lab-odpi-egeria-lab-core-0 in StatefulSet lab-odpi-egeria-lab-core success
     ```
     
     this therefore suggests all should work, but until the pvc claim is resolved nothing will start....

**Comment by planetf1** [comment]
(aside: some of the resource names could do with tidying up)

**Comment by planetf1** [comment]
Probably useful to:
 * Allow the storageClass to be set
 * Add info on persistence to the chart readme
 * Allow the storage size to be set
 
 More generally whilst the lab chart is useful, a real deployment will likely use their own charts. We may want to create a more appropriate/simpler one, and/or document more thoroughly how to write them.
 
 Also note - no persistence for the jupyter notebook currently. (and not needed for UIs)

**Comment by planetf1** [comment]
Due to an issue with my cluster (!) I needed to fix up the environment (applied update, which restarted all workers, as the storage allocation pod was not working!). The end result is pvcs now have real storage and pods have started ie:
```
$ kubectl get pvc                                                                                           [8:18:43]
NAME                                               STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                 AGE
data-lab-kafka-0                                   Bound    pvc-e92acd6b-f5db-41ea-80b3-6ee0d3283f1d   10Gi       RWO            ibmc-vpc-block-10iops-tier   13h
data-lab-zookeeper-0                               Bound    pvc-e97b60c5-ef3d-4077-aea9-a81f27147a36   10Gi       RWO            ibmc-vpc-block-10iops-tier   13h
lab-core-data-lab-odpi-egeria-lab-core-0           Bound    pvc-adc76b97-c6e6-49b0-9ac0-a9fbbcd468a5   10Gi       RWO            ibmc-vpc-block-10iops-tier   13h
lab-datalake-data-lab-odpi-egeria-lab-datalake-0   Bound    pvc-a182e29a-3dca-40d8-8e16-823ab0338b77   10Gi       RWO            ibmc-vpc-block-10iops-tier   13h
lab-dev-data-lab-odpi-egeria-lab-dev-0             Bound    pvc-1934bfe9-d324-424c-8939-03b09f5020c1   10Gi       RWO            ibmc-vpc-block-10iops-tier   13h
lab-factory-data-lab-odpi-egeria-lab-factory-0     Bound    pvc-26e31013-1146-4d72-ad4e-9db6b9206851   10Gi       RWO            ibmc-vpc-block-10iops-tier   13h
```
```
$ kubectl get pv                                                                                                                                [8:20:26]
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                                         STORAGECLASS                 REASON   AGE
pvc-1934bfe9-d324-424c-8939-03b09f5020c1   10Gi       RWO            Delete           Bound    egeria-lab/lab-dev-data-lab-odpi-egeria-lab-dev-0             ibmc-vpc-block-10iops-tier            13h
pvc-26e31013-1146-4d72-ad4e-9db6b9206851   10Gi       RWO            Delete           Bound    egeria-lab/lab-factory-data-lab-odpi-egeria-lab-factory-0     ibmc-vpc-block-10iops-tier            13h
pvc-a182e29a-3dca-40d8-8e16-823ab0338b77   10Gi       RWO            Delete           Bound    egeria-lab/lab-datalake-data-lab-odpi-egeria-lab-datalake-0   ibmc-vpc-block-10iops-tier            13h
pvc-adc76b97-c6e6-49b0-9ac0-a9fbbcd468a5   10Gi       RWO            Delete           Bound    egeria-lab/lab-core-data-lab-odpi-egeria-lab-core-0           ibmc-vpc-block-10iops-tier            13h
pvc-e92acd6b-f5db-41ea-80b3-6ee0d3283f1d   10Gi       RWO            Delete           Bound    egeria-lab/data-lab-kafka-0                                   ibmc-vpc-block-10iops-tier            13h
pvc-e97b60c5-ef3d-4077-aea9-a81f27147a36   10Gi       RWO            Delete           Bound    egeria-lab/data-lab-zookeeper-0                               ibmc-vpc-block-10iops-tier            13h
```

**Comment by planetf1** [comment]
When the config notebook is run we see:
```
Configuring cocoMDS1...
   ... configuring the platform the server will run on...
   ...... (POST https://lab-dev:9443/open-metadata/admin-services/users/garygeeke/servers/cocoMDS1/server-url-root?url=https://lab-datalake:9443 )
   ...... Response:  {'class': 'VoidResponse', 'relatedHTTPCode': 500, 'exceptionClassName': 'org.odpi.openmetadata.adminservices.ffdc.exception.OMAGConfigurationErrorException', 'exceptionCausedBy': 'org.odpi.openmetadata.frameworks.connectors.ffdc.OCFRuntimeException', 'actionDescription': 'setServerURLRoot', 'exceptionErrorMessage': "OMAG-ADMIN-500-001 Method setServerURLRoot for OMAG server cocoMDS1 returned an unexpected exception of org.odpi.openmetadata.frameworks.connectors.ffdc.OCFRuntimeException with message ENCRYPTED-DOC-STORE-400-002  Unable to write encrypted configuration file; exception was java.io.IOException with message Directory 'data/servers/cocoMDS1/config' could not be created", 'exceptionErrorMessageId': 'OMAG-ADMIN-500-001', 'exceptionErrorMessageParameters': ['cocoMDS1', 'setServerURLRoot', 'org.odpi.openmetadata.frameworks.connectors.ffdc.OCFRuntimeException', "ENCRYPTED-DOC-STORE-400-002  Unable to write encrypted configuration file; exception was java.io.IOException with message Directory 'data/servers/cocoMDS1/config' could not be created"], 'exceptionSystemAction': "The system is unable to configure the OMAG server.  No change was made to the server's configuration document.", 'exceptionUserAction': 'This is likely to be either an operational or logic error. Look for other errors.  Validate the request.  If you are stuck, raise an issue.'}
   ```
   Looking within the image we can see a data directory. however the user running egeria has no write permission ....
   ```
   [jboss@lab-odpi-egeria-lab-core-0 ~]$ pwd
/home/jboss
[jboss@lab-odpi-egeria-lab-core-0 ~]$ cd /deployments/
[jboss@lab-odpi-egeria-lab-core-0 deployments]$ ls
LICENSE  NOTICE  conformance-suite  data  keystore_0dEJAOTu_wBrY_yE6nICNer0Q4aPcxhu  samples  server  truststore.p12  user-interface  utilities
[jboss@lab-odpi-egeria-lab-core-0 deployments]$ ls -la
total 208
drwxrwxr-x. 1 jboss root   4096 Dec 18 08:59 .
drwxr-xr-x. 1 root  root   4096 Dec 17 19:15 ..
-rw-rw-r--. 1 root  root  11483 Dec 17 15:53 LICENSE
-rwxrwxr-x. 1 root  root    268 Dec 17 15:53 NOTICE
drwxrwxr-x. 2 root  root   4096 Dec 17 15:53 conformance-suite
drwxr-xr-x. 3 root  root   4096 Dec 17 19:14 data
drwx------. 2 jboss root   4096 Dec 18 08:59 keystore_0dEJAOTu_wBrY_yE6nICNer0Q4aPcxhu
drwxrwxr-x. 2 root  root   4096 Dec 17 15:53 samples
drwxrwxr-x. 3 root  root   4096 Dec 17 15:53 server
-rw-rw-r--. 1 root  root 119761 Dec 17 15:53 truststore.p12
drwxrwxr-x. 2 root  root   4096 Dec 17 15:53 user-interface
drwxrwxr-x. 2 root  root   4096 Dec 17 15:53 utilities
[jboss@lab-odpi-egeria-lab-core-0 deployments]$ cd data
[jboss@lab-odpi-egeria-lab-core-0 data]$ ls
lost+found
[jboss@lab-odpi-egeria-lab-core-0 data]$ ls -la
total 28
drwxr-xr-x. 3 root  root  4096 Dec 17 19:14 .
drwxrwxr-x. 1 jboss root  4096 Dec 18 08:59 ..
drwx------. 2 root  root 16384 Dec 17 19:14 lost+found
[jboss@lab-odpi-egeria-lab-core-0 data]$ touch file
touch: cannot touch 'file': Permission denied
```

So permissions need correcting ... (this is specific to the k8s case)

**Comment by planetf1** [comment]
The openshift docs at https://docs.openshift.com/container-platform/3.5/install_config/persistent_storage/pod_security_context.html
say
```

Allowing access to SCCs with a RunAsAny FSGroup strategy can also prevent users from accessing their block devices. Pods need to specify an fsGroup in order to take over their block devices. Normally, this is done when the SCC FSGroup strategy is set to MustRunAs. If a user’s pod is assigned an SCC with a RunAsAny FSGroup strategy, then the user may face permission denied errors until they discover that they need to specify an fsGroup themselves.
--



```
One might expect similar constraints on other enterprise k8s deployments.

**Comment by planetf1** [comment]
After adding a 'securityContext.fsGroup:0' setting (this says which unix group is used for storage.. 0 (root - but only as a group) is what the redhat ubi8 jdk image uses as default), storage is correctly accessible by each pod.

I deployed, configured, then killed a pod. It restarted, and after restart I could see /deployments/data still contains the correct configuration etc files we would expect.

it's worth noting that pod rescheduling is a common experience. But currently with egeria the on-disk state is not enough to get the platform back up and running - the servers need to be started - or configuration set to autostart.

This is one area where an 'operator' will be able to do a slightly more thorough job - it could monitor the pods readiness then issue the right calls to start the configured servers.

However after restart now seeing that the notebook start (in a restarted pod ie platform) is returning notebook errors. investigating:

Checking OMAG Server Platform availability...
    Core Platform is active
        Checking OMAG Server cocoMDS2
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/common/environment-check.ipynb in <module>
      6 print("\nChecking OMAG Server Platform availability...")
      7 
----> 8 activatePlatform(corePlatformName, corePlatformURL, [cocoMDS2Name, cocoMDS3Name, cocoMDS5Name, cocoMDS6Name])
      9 activatePlatform(dataLakePlatformName, dataLakePlatformURL, [cocoMDS1Name, cocoMDS4Name, cocoView1Name ])
     10 activatePlatform(devPlatformName, devPlatformURL, [cocoMDSxName])

**Comment by planetf1** [comment]
The cause is most likely:
```
$ http --verbose --verify=no --pretty=format GET https://a32e304b-eu-gb.lb.appdomain.cloud:9443/open-metadata/admin-services/users/garygeeke/servers/cocoMDS2/configuration   [14:40:01]
GET /open-metadata/admin-services/users/garygeeke/servers/cocoMDS2/configuration HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: a32e304b-eu-gb.lb.appdomain.cloud:9443
User-Agent: HTTPie/2.3.0



HTTP/1.1 200
Connection: keep-alive
Content-Type: application/json
Date: Fri, 18 Dec 2020 14:40:39 GMT
Keep-Alive: timeout=60
Transfer-Encoding: chunked

{
    "actionDescription": "getStoredConfiguration",
    "class": "OMAGServerConfigResponse",
    "exceptionCausedBy": "org.odpi.openmetadata.frameworks.connectors.ffdc.OCFRuntimeException",
    "exceptionClassName": "org.odpi.openmetadata.adminservices.ffdc.exception.OMAGConfigurationErrorException",
    "exceptionErrorMessage": "OMAG-ADMIN-500-001 Method getStoredConfiguration for OMAG server cocoMDS2 returned an unexpected exception of org.odpi.openmetadata.frameworks.connectors.ffdc.OCFRuntimeException with message ENCRYPTED-DOC-STORE-400-011  Unable to find any keys defined for decrypting the configuration.",
    "exceptionErrorMessageId": "OMAG-ADMIN-500-001",
    "exceptionErrorMessageParameters": [
        "cocoMDS2",
        "getStoredConfiguration",
        "org.odpi.openmetadata.frameworks.connectors.ffdc.OCFRuntimeException",
        "ENCRYPTED-DOC-STORE-400-011  Unable to find any keys defined for decrypting the configuration."
    ],
    "exceptionSystemAction": "The system is unable to configure the OMAG server.  No change was made to the server's configuration document.",
    "exceptionUserAction": "This is likely to be either an operational or logic error. Look for other errors.  Validate the request.  If you are stuck, raise an issue.",
    "relatedHTTPCode": 500
}
```
This may be an issue with the change to file paths - need to check where config doc key is stored

**Comment by planetf1** [comment]
See https://egeria-project.org/open-metadata-implementation/adapters/open-connectors/configuration-store-connectors/configuration-encrypted-file-store-connector/. Need to figure out best way to set EGERIA_CONFIG_KEYS. This is seemingly being written to the fs on first start - but if we moved that
 a) it would be more consistent -- always write data under /data
 but
 b)  it would make little real sense for security if the key is co-located with the data
 so
 c) This is best managed in a k8s secret in that environment
 
 Inclination is to make the file change as a quick fix (in part as we also want this to work with docker alone, and with docker-compose), then return and add a k8s secret later.

**Comment by planetf1** [comment]
Worked through bugs.

Lab helm chart now works with persistence by default.

A later PR will be used to extend the docs & add configuration of storage class.

**Comment by planetf1** [comment]
Changing default key location to data/platform/keys
Added to release notes

**Comment by planetf1** [comment]
Sufficient work has now been done to enable this capability & support in the standard helm chart.

See #4454 where a new simple chart will be created for deployment, and support persistance. This will be a better base for experimentation beyond a tutorial, and is more likely the place where persistance is wanted
